### PR TITLE
Add Pre commit

### LINKED ISSUE #74: Add pre-commit
Pre-commits are good, we should have some. It would include `black` and ruff, and the custom notebook hook we had.

This is good for us and can also serve as an instructive example for any students who look at the repo.

### PR DESCRIPTION
**Pull request by AlinderS**
This PR adds pre-commit to the Pipfile and a config file for pre-commit containing
- ruff
- black
- codespell
- nbstripout
- check-yaml

This  will resolve #74 

I will add another hook that checks for and deletes unused images and I will add a commit that fixes all the issues these new hooks raise.

### CODE REVIEW

File: .pre-commit-config.yaml [hunk 1]
```diff
@@ -0,0 +1,31 @@
+repos:
+-   repo: https://github.com/pre-commit/pre-commit-hooks
+    rev: v4.5.0
+    hooks:
+    -   id: check-yaml
```
Comment (eerovaher) [inline .pre-commit-config.yaml:5]: There's a few more hooks that would be good to enable:
```suggestion
      - id: check-added-large-files
      - id: check-merge-conflict
      - id: check-yaml
      - id: detect-private-key
      - id: end-of-file-fixer
      - id: trailing-whitespace
```

**Comment by AlinderS** [comment]
> Have you thought what would happen if #65 was merged first and this pull request after that?

Yes. If #65 is merged first, this PR will be automatically rebased to master. GitHub handles it automatically.

File: .pre-commit-config.yaml (hunk 1 above)
Comment (AlinderS) [reply .pre-commit-config.yaml:5]: Are not `trailing-whitespace` covered by [Ruff rule W291](https://docs.astral.sh/ruff/rules/trailing-whitespace/) and `end-of-file-fixer` by [Ruff rule W292](https://docs.astral.sh/ruff/rules/missing-newline-at-end-of-file/)?

File: 4-tools/manual.ipynb [hunk 2]
```diff
@@ -34,7 +34,7 @@
     "[`black`](https://black.readthedocs.io/en/stable/index.html) is a commonly used and very opinionated automatic code formatter. \n",
     "It takes the burden of hand-formatting away from the user and lets them focus on the content. \n",
     "Clearly formatted code is easier and faster to read. `black` has hardly any settings and instead enforces good defaults. \n",
-    "This makes `black`ened code consistent across projects. \n",
+    "This makes `black` code consistent across projects. \n",
```
[the remark this answers is not in the document]
Comment (eerovaher) [reply 4-tools/manual.ipynb:37]: If "blackened" is good enough for Black itself then it's good enough for us too.

File: .pre-commit-config.yaml (hunk 1 above)
Comment (eerovaher) [reply .pre-commit-config.yaml:5]: Ruff is a Python linter, it doesn't check all files.

**Pull request merged**